// @glimmer/vm-babel-plugins/dist/dev/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var node_module = require('node:module');

const require$1 = node_module.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (document.currentScript && document.currentScript.src || new URL('index.cjs', document.baseURI).href)));
function defaultTo(value, defaultVal) {
  return value === undefined ? defaultVal : value;
}
function generateVmPlugins() {
  let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let isDebug = defaultTo(options.isDebug, true);
  let __loadPlugins = defaultTo(options.__loadPlugins, false);
  return [[__loadPlugins ? require$1('babel-plugin-debug-macros') : require$1.resolve('babel-plugin-debug-macros'), {
    debugTools: {
      source: '@glimmer/global-context',
      isDebug
    },
    externalizeHelpers: {
      module: true
    },
    flags: [{
      source: '@glimmer/env',
      flags: {
        DEBUG: isDebug
      }
    }]
  }, 'glimmer-vm-debug-macros']];
}

exports.default = generateVmPlugins;
