// @glimmer/program/dist/dev/index.cjs
'use strict';

var manager = require('@glimmer/manager');
var opcodeCompiler = require('@glimmer/opcode-compiler');
var util = require('@glimmer/util');
var vm = require('@glimmer/vm');
var wireFormat = require('@glimmer/wire-format');

function _typeof(obj) {
  "@babel/helpers - typeof";

  return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) {
    return typeof obj;
  } : function (obj) {
    return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
  }, _typeof(obj);
}

function _toPrimitive(input, hint) {
  if (_typeof(input) !== "object" || input === null) return input;
  var prim = input[Symbol.toPrimitive];
  if (prim !== undefined) {
    var res = prim.call(input, hint || "default");
    if (_typeof(res) !== "object") return res;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (hint === "string" ? String : Number)(input);
}

function _toPropertyKey(arg) {
  var key = _toPrimitive(arg, "string");
  return _typeof(key) === "symbol" ? key : String(key);
}

function _defineProperty(obj, key, value) {
  key = _toPropertyKey(key);
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }
  return obj;
}

/**
 * Default component template, which is a plain yield
 */
const DEFAULT_TEMPLATE_BLOCK = [[[wireFormat.SexpOpcodes.Yield, 1, null]], ['&default'], false, []];
const DEFAULT_TEMPLATE = {
  // random uuid
  id: '1b32f5c2-7623-43d6-a0ad-9672898920a1',
  moduleName: '__default__.hbs',
  block: JSON.stringify(DEFAULT_TEMPLATE_BLOCK),
  scope: null,
  isStrictMode: true
};

const WELL_KNOWN_EMPTY_ARRAY = Object.freeze([]);
const STARTER_CONSTANTS = util.constants(WELL_KNOWN_EMPTY_ARRAY);
const WELL_KNOWN_EMPTY_ARRAY_POSITION = STARTER_CONSTANTS.indexOf(WELL_KNOWN_EMPTY_ARRAY);
class CompileTimeConstantImpl {
  constructor() {
    // `0` means NULL
    _defineProperty(this, "values", STARTER_CONSTANTS.slice());
    _defineProperty(this, "indexMap", new Map(this.values.map((value, index) => [value, index])));
  }
  value(value) {
    let indexMap = this.indexMap;
    let index = indexMap.get(value);
    if (index === undefined) {
      index = this.values.push(value) - 1;
      indexMap.set(value, index);
    }
    return index;
  }
  array(values) {
    if (values.length === 0) {
      return WELL_KNOWN_EMPTY_ARRAY_POSITION;
    }
    let handles = new Array(values.length);
    for (let i = 0; i < values.length; i++) {
      handles[i] = this.value(values[i]);
    }
    return this.value(handles);
  }
  toPool() {
    return this.values;
  }
}
class RuntimeConstantsImpl {
  constructor(pool) {
    _defineProperty(this, "values", void 0);
    this.values = pool;
  }
  getValue(handle) {
    return this.values[handle];
  }
  getArray(value) {
    let handles = this.getValue(value);
    let reified = new Array(handles.length);
    for (const [i, n] of util.enumerate(handles)) {
      reified[i] = this.getValue(n);
    }
    return reified;
  }
}
class ConstantsImpl extends CompileTimeConstantImpl {
  constructor() {
    super(...arguments);
    _defineProperty(this, "reifiedArrs", {
      [WELL_KNOWN_EMPTY_ARRAY_POSITION]: WELL_KNOWN_EMPTY_ARRAY
    });
    _defineProperty(this, "defaultTemplate", opcodeCompiler.templateFactory(DEFAULT_TEMPLATE)());
    // Used for tests and debugging purposes, and to be able to analyze large apps
    // This is why it's enabled even in production
    _defineProperty(this, "helperDefinitionCount", 0);
    _defineProperty(this, "modifierDefinitionCount", 0);
    _defineProperty(this, "componentDefinitionCount", 0);
    _defineProperty(this, "helperDefinitionCache", new WeakMap());
    _defineProperty(this, "modifierDefinitionCache", new WeakMap());
    _defineProperty(this, "componentDefinitionCache", new WeakMap());
  }
  helper(definitionState) {
    let isOptional = arguments.length > 2 ? arguments[2] : undefined;
    let handle = this.helperDefinitionCache.get(definitionState);
    if (handle === undefined) {
      let managerOrHelper = manager.getInternalHelperManager(definitionState, isOptional);
      if (managerOrHelper === null) {
        this.helperDefinitionCache.set(definitionState, null);
        return null;
      }
      util.assert(managerOrHelper, 'BUG: expected manager or helper');
      let helper = typeof managerOrHelper === 'function' ? managerOrHelper : managerOrHelper.getHelper(definitionState);
      handle = this.value(helper);
      this.helperDefinitionCache.set(definitionState, handle);
      this.helperDefinitionCount++;
    }
    return handle;
  }
  modifier(definitionState) {
    let resolvedName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
    let isOptional = arguments.length > 2 ? arguments[2] : undefined;
    let handle = this.modifierDefinitionCache.get(definitionState);
    if (handle === undefined) {
      let manager$1 = manager.getInternalModifierManager(definitionState, isOptional);
      if (manager$1 === null) {
        this.modifierDefinitionCache.set(definitionState, null);
        return null;
      }
      let definition = {
        resolvedName,
        manager: manager$1,
        state: definitionState
      };
      handle = this.value(definition);
      this.modifierDefinitionCache.set(definitionState, handle);
      this.modifierDefinitionCount++;
    }
    return handle;
  }
  component(definitionState, owner, isOptional) {
    let definition = this.componentDefinitionCache.get(definitionState);
    if (definition === undefined) {
      let manager$1 = manager.getInternalComponentManager(definitionState, isOptional);
      if (manager$1 === null) {
        this.componentDefinitionCache.set(definitionState, null);
        return null;
      }
      util.assert(manager$1, 'BUG: expected manager');
      let capabilities = manager.capabilityFlagsFrom(manager$1.getCapabilities(definitionState));
      let templateFactory = manager.getComponentTemplate(definitionState);
      let compilable = null;
      let template;
      if (!manager.managerHasCapability(manager$1, capabilities, vm.InternalComponentCapabilities.dynamicLayout)) {
        template = (templateFactory === null || templateFactory === void 0 ? void 0 : templateFactory(owner)) ?? this.defaultTemplate;
      } else {
        template = templateFactory === null || templateFactory === void 0 ? void 0 : templateFactory(owner);
      }
      if (template !== undefined) {
        template = util.unwrapTemplate(template);
        compilable = manager.managerHasCapability(manager$1, capabilities, vm.InternalComponentCapabilities.wrapped) ? template.asWrappedLayout() : template.asLayout();
      }
      definition = {
        resolvedName: null,
        handle: -1,
        // replaced momentarily
        manager: manager$1,
        capabilities,
        state: definitionState,
        compilable
      };
      definition.handle = this.value(definition);
      this.componentDefinitionCache.set(definitionState, definition);
      this.componentDefinitionCount++;
    }
    return definition;
  }
  resolvedComponent(resolvedDefinition, resolvedName) {
    let definition = this.componentDefinitionCache.get(resolvedDefinition);
    if (definition === undefined) {
      let {
        manager: manager$1,
        state,
        template
      } = resolvedDefinition;
      let capabilities = manager.capabilityFlagsFrom(manager$1.getCapabilities(resolvedDefinition));
      let compilable = null;
      if (!manager.managerHasCapability(manager$1, capabilities, vm.InternalComponentCapabilities.dynamicLayout)) {
        template = template ?? this.defaultTemplate;
      }
      if (template !== null) {
        template = util.unwrapTemplate(template);
        compilable = manager.managerHasCapability(manager$1, capabilities, vm.InternalComponentCapabilities.wrapped) ? template.asWrappedLayout() : template.asLayout();
      }
      definition = {
        resolvedName,
        handle: -1,
        // replaced momentarily
        manager: manager$1,
        capabilities,
        state,
        compilable
      };
      definition.handle = this.value(definition);
      this.componentDefinitionCache.set(resolvedDefinition, definition);
      this.componentDefinitionCount++;
    }
    return util.expect(definition, 'BUG: resolved component definitions cannot be null');
  }
  getValue(index) {
    util.assert(index >= 0, `cannot get value for handle: ${index}`);
    return this.values[index];
  }
  getArray(index) {
    let reifiedArrs = this.reifiedArrs;
    let reified = reifiedArrs[index];
    if (reified === undefined) {
      let names = this.getValue(index);
      reified = new Array(names.length);
      for (const [i, name] of util.enumerate(names)) {
        reified[i] = this.getValue(name);
      }
      reifiedArrs[index] = reified;
    }
    return reified;
  }
}

class RuntimeOpImpl {
  constructor(heap) {
    _defineProperty(this, "offset", 0);
    this.heap = heap;
  }
  get size() {
    let rawType = this.heap.getbyaddr(this.offset);
    return ((rawType & vm.OPERAND_LEN_MASK) >> vm.ARG_SHIFT) + 1;
  }
  get isMachine() {
    let rawType = this.heap.getbyaddr(this.offset);
    return rawType & vm.MACHINE_MASK ? 1 : 0;
  }
  get type() {
    return this.heap.getbyaddr(this.offset) & vm.TYPE_MASK;
  }
  get op1() {
    return this.heap.getbyaddr(this.offset + 1);
  }
  get op2() {
    return this.heap.getbyaddr(this.offset + 2);
  }
  get op3() {
    return this.heap.getbyaddr(this.offset + 3);
  }
}

var TableSlotState = /*#__PURE__*/function (TableSlotState) {
  TableSlotState[TableSlotState["Allocated"] = 0] = "Allocated";
  TableSlotState[TableSlotState["Freed"] = 1] = "Freed";
  TableSlotState[TableSlotState["Purged"] = 2] = "Purged";
  TableSlotState[TableSlotState["Pointer"] = 3] = "Pointer";
  return TableSlotState;
}(TableSlotState || {});
const PAGE_SIZE = 0x100000;
class RuntimeHeapImpl {
  constructor(serializedHeap) {
    _defineProperty(this, "heap", void 0);
    _defineProperty(this, "table", void 0);
    let {
      buffer,
      table
    } = serializedHeap;
    this.heap = new Int32Array(buffer);
    this.table = table;
  }

  // It is illegal to close over this address, as compaction
  // may move it. However, it is legal to use this address
  // multiple times between compactions.
  getaddr(handle) {
    return util.unwrap(this.table[handle]);
  }
  getbyaddr(address) {
    return util.expect(this.heap[address], 'Access memory out of bounds of the heap');
  }
  sizeof(handle) {
    return sizeof(this.table);
  }
}
function hydrateHeap(serializedHeap) {
  return new RuntimeHeapImpl(serializedHeap);
}

/**
 * The Heap is responsible for dynamically allocating
 * memory in which we read/write the VM's instructions
 * from/to. When we malloc we pass out a VMHandle, which
 * is used as an indirect way of accessing the memory during
 * execution of the VM. Internally we track the different
 * regions of the memory in an int array known as the table.
 *
 * The table 32-bit aligned and has the following layout:
 *
 * | ... | hp (u32) |       info (u32)   | size (u32) |
 * | ... |  Handle  | Scope Size | State | Size       |
 * | ... | 32bits   | 30bits     | 2bits | 32bit      |
 *
 * With this information we effectively have the ability to
 * control when we want to free memory. That being said you
 * can not free during execution as raw address are only
 * valid during the execution. This means you cannot close
 * over them as you will have a bad memory access exception.
 */
class HeapImpl {
  constructor() {
    _defineProperty(this, "offset", 0);
    _defineProperty(this, "heap", void 0);
    _defineProperty(this, "handleTable", void 0);
    _defineProperty(this, "handleState", void 0);
    _defineProperty(this, "handle", 0);
    this.heap = new Int32Array(PAGE_SIZE);
    this.handleTable = [];
    this.handleState = [];
  }
  pushRaw(value) {
    this.sizeCheck();
    this.heap[this.offset++] = value;
  }
  pushOp(item) {
    this.pushRaw(item);
  }
  pushMachine(item) {
    this.pushRaw(item | vm.MACHINE_MASK);
  }
  sizeCheck() {
    let {
      heap
    } = this;
    if (this.offset === this.heap.length) {
      let newHeap = new Int32Array(heap.length + PAGE_SIZE);
      newHeap.set(heap, 0);
      this.heap = newHeap;
    }
  }
  getbyaddr(address) {
    return util.unwrap(this.heap[address]);
  }
  setbyaddr(address, value) {
    this.heap[address] = value;
  }
  malloc() {
    // push offset, info, size
    this.handleTable.push(this.offset);
    return this.handleTable.length - 1;
  }
  finishMalloc(handle) {
  }
  size() {
    return this.offset;
  }

  // It is illegal to close over this address, as compaction
  // may move it. However, it is legal to use this address
  // multiple times between compactions.
  getaddr(handle) {
    return util.unwrap(this.handleTable[handle]);
  }
  sizeof(handle) {
    return sizeof(this.handleTable);
  }
  free(handle) {
    this.handleState[handle] = TableSlotState.Freed;
  }

  /**
   * The heap uses the [Mark-Compact Algorithm](https://en.wikipedia.org/wiki/Mark-compact_algorithm) to shift
   * reachable memory to the bottom of the heap and freeable
   * memory to the top of the heap. When we have shifted all
   * the reachable memory to the top of the heap, we move the
   * offset to the next free position.
   */
  compact() {
    let compactedSize = 0;
    let {
      handleTable,
      handleState,
      heap
    } = this;
    for (let i = 0; i < length; i++) {
      let offset = util.unwrap(handleTable[i]);
      let size = util.unwrap(handleTable[i + 1]) - util.unwrap(offset);
      let state = handleState[i];
      if (state === TableSlotState.Purged) {
        continue;
      } else if (state === TableSlotState.Freed) {
        // transition to "already freed" aka "purged"
        // a good improvement would be to reuse
        // these slots
        handleState[i] = TableSlotState.Purged;
        compactedSize += size;
      } else if (state === TableSlotState.Allocated) {
        for (let j = offset; j <= i + size; j++) {
          heap[j - compactedSize] = util.unwrap(heap[j]);
        }
        handleTable[i] = offset - compactedSize;
      } else if (state === TableSlotState.Pointer) {
        handleTable[i] = offset - compactedSize;
      }
    }
    this.offset = this.offset - compactedSize;
  }
  capture() {
    let offset = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.offset;
    // Only called in eager mode
    let buffer = slice(this.heap, 0, offset).buffer;
    return {
      handle: this.handle,
      table: this.handleTable,
      buffer: buffer
    };
  }
}
class RuntimeProgramImpl {
  constructor(constants, heap) {
    _defineProperty(this, "_opcode", void 0);
    this.constants = constants;
    this.heap = heap;
    this._opcode = new RuntimeOpImpl(this.heap);
  }
  opcode(offset) {
    this._opcode.offset = offset;
    return this._opcode;
  }
}
function slice(arr, start, end) {
  if (arr.slice !== undefined) {
    return arr.slice(start, end);
  }
  let ret = new Int32Array(end);
  for (; start < end; start++) {
    ret[start] = util.unwrap(arr[start]);
  }
  return ret;
}
function sizeof(table, handle) {
  {
    return -1;
  }
}

function artifacts() {
  return {
    constants: new ConstantsImpl(),
    heap: new HeapImpl()
  };
}

exports.CompileTimeConstantImpl = CompileTimeConstantImpl;
exports.ConstantsImpl = ConstantsImpl;
exports.HeapImpl = HeapImpl;
exports.RuntimeConstantsImpl = RuntimeConstantsImpl;
exports.RuntimeHeapImpl = RuntimeHeapImpl;
exports.RuntimeOpImpl = RuntimeOpImpl;
exports.RuntimeProgramImpl = RuntimeProgramImpl;
exports.artifacts = artifacts;
exports.hydrateHeap = hydrateHeap;
